### DISCUSSION TITLE
Renovate cache is growing indefinitely since 2024-06-26

### DISCUSSION THREAD

**Discussion by oliver-goetz** [renovatebot/renovate#30525] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitHub, latest version of Renovate (auto-update)

### Please tell us more about your question or problem

Since 2024-06-26 the cache of our self hosted renovate instance is growing indefinitely. Especially the growing number of inodes is critical.
![Screenshot 2024-08-01 at 12 06 59](https://github.com/user-attachments/assets/7137a43f-57a0-4420-b02b-b2a231e20860)
On 2024-07-08 I increased the disk size and on 2024-07-18 I deleted the cache disk. These are the reasons for the spikes in the diagram. Otherwise, you see a constant grow of the the cache volume size and number of inodes.

This behavior starts the day https://github.com/renovatebot/renovate/pull/29860 was merged. We auto update renovate, so the refactoring of the cache cleanup is probably related.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by RahulGautamSingh** [comment]
Thanks.
 
It seems like a `cacache.rm.verify` call is needed from time-to-time, even though we are removing both entry and content. We removed verify because it was taking too much time for some users, if called on each run.

I have 2 potential solutions in mind for this:
1. try `rm.entry` with `opts.removeFully` + `rm.content` -- maybe the indexes keep growing as `rm.entry` only added `null` & does no cleanup of old entries.

2. use `rm.verify.lastRun` + `rmverify` -> to run the `verify` at certain time intervals rather than each run.

I will try them locally and see which works better.

**Reply by rarkins** [reply]
Can this be demonstrated at a small scale?

**Reply by RahulGautamSingh** [reply]
That's the issue while testing locally.

**Reply by rarkins** [reply]
So we can't have a small cache, wait for some to expire, then check if they were properly flushed?

**Reply by RahulGautamSingh** [reply]
Yes, I was gonna try this only..set expiry to seconds and check if all are removed. The issue is that I checked this the last time and it worked.

**Reply by rarkins** [reply]
@oliver-goetz Can you try to give steps to reproduce locally?

**Reply by oliver-goetz** [reply]
I don't exactly know how to reproduce it locally, because I don't exactly what I should check. I just observed this behavior while running renovate in our Kubernetes cluster.
However, I can share our configuration. You might be able to reproduce the issue with that.
https://github.com/gardener/ci-infra/blob/master/config/prow/cluster/renovate/renovate_deployment.yaml

**Reply by rarkins** [reply]
We'll need to wait until someone can reproduce locally

**Reply by tibeer** [reply]
My self hosted instance is taking care of just a few things and the 1Gi cache exceeded this night. So it might be coincidence, might be the same issue. I'm running `renovate:37.440.7` and the job itself runs every 5min. I cleared the whole cache (`rm -rf`) and am now at 19432KB storage consumtion after clearing and first run. Will keep an eye on my cache and update here if I notice unusual increments.

**Reply by tibeer** [reply]
3 days later an the storage consumption now looks like this:

```sh
root@debian:/debug# du -sc ./*
344516	./cache
16	./lost+found
12	./repos
344544	total
```

Will still keep an eye on it.

**Reply by tibeer** [reply]
Another 3 days later:

```sh
root@debian:/debug# du -sc ./*
575008	./cache
16	./lost+found
12	./repos
575036	total
```

**Reply by tibeer** [reply]
Due to vacation I was not able to monitor this with more detail. But after returning I noticed that the container is no longer starting because of a full disk. So jeah, I can confirm that the storage consumption is constantly increasing and cleanup isn't working as expected.

**Reply by rarkins** [reply]
Unfortunately it still needs someone to provide reproducible steps at a small scale which can be tested against

**Reply by viceice** [reply]
try sqlite or redis cache, they need much less dist space and are much faster

**Reply by tibeer** [reply]
Finally managed to continue on this. I switched now to redis and cleaned the disk again:
```
root@debian:/debug# du -sc ./*
22516	./cache
12	./repos
22528	total
```
I will report back if the disk usage is drastically increasing again or if it working now.

**Reply by tibeer** [reply]
A bunch of PRs from renovate bot later, the disk usage still looks like this:

```sh
root@debian:/debug# du -sc ./*
22516	./cache
12	./repos
22528	total
```

I guess using the file system for caching is pretty suboptimal. Looking into the redis storage consumption, this looks much smaller compared to the file system:

```sh
I have no name!@renovate-redis-master-0:/data$ du -sch ./*
9.5M	./appendonlydir
16K	./lost+found
9.5M	total
```

I made a PR to extend the charts values with a comment that recommends redis if one wants to use the cache:
https://github.com/renovatebot/helm-charts/pull/1799

**Reply by tibeer** [reply]
Again a few days later, this is how the storage consumption now looks on the redis node:

```sh
I have no name!@renovate-redis-master-0:/data$ du -sch ./*
18M	./appendonlydir
16K	./lost+found
18M	total
```

The consumption of the job pod has not changed since then. I guess this ticket can be closed then @oliver-goetz ?

**Reply by oliver-goetz** [reply]
From my point of view the disk cache feature is still broken. Why don't you revert https://github.com/renovatebot/renovate/pull/29860 which is causing the issue?
If you recommend using redis for disk cache anyway, it would be better having an slower but unbroken disk cache feature instead of a faster broken one.

**Reply by rarkins** [reply]
We don't roll back because the existing approach of using `cacache.rm.entry()` also had a problem with clearing disk space: https://github.com/renovatebot/renovate/discussions/28169

As far as we can tell, Renovate's use of `cacache` has _never_ cleared up disk space correctly, so there is no "regression" from that perspective.

Have you tried switching to sqlite for the local cache?

**Reply by lindhe** [reply]
How would one enable `sqlite`? 🤔

**Reply by jamietanna** [reply]
From a cursory search of `sqlite` in our docs, there's https://docs.renovatebot.com/self-hosted-experimental/#renovate_x_sqlite_package_cache for the package cache

**Reply by lindhe** [reply]
Thanks. I found that, but I thought it was wrong since they are experimental... 😕 There's a comment in the Values file indicating that using SQLite is encouraged:

https://github.com/renovatebot/helm-charts/blob/fbfc69ab9c82c720d867f92d38c27d75d8c902e6/charts/renovate/values.yaml#L127-L128

Specifically, I was dissuaded by the text on that docs page you linked:

> The following environment variables are "experimental" because they:
> 
> - are not commonly needed
> - are typically an effort to work around some other service's or platform's problem
> - can be removed at any time
> - are variables for Renovate's internal use to validate they work as intended 
> 
> Experimental variables which are commonly used and for which there is no external solution in sight can be converted to an official configuration option by the Renovate bot developers.
> 
> Use these experimental variables at your own risk. We do not follow Semantic Versioning for any experimental variables. These variables may be removed or have their behavior changed in any version. We will try to keep breakage to a minimum, but make no guarantees that an experimental variable will keep working.

If file system cache is not reliable and SQLite is experimental, is there no reliable way to enable cache? 😕

**Reply by jamietanna** [reply]
Ah interesting - I'd wondered why the recent nudge on this old thread :) 

Using experimental variables are fine, and (without checking) my gut feel is we don't break them that often - but it is always a "be careful"

I'll see internally if we're confident in its stability to being potentially promoted to a non-experimental option

**Reply by lindhe** [reply]
Thanks a lot for the input. Much appreciated! 🫡
